pkgs/build-support/gradle/update-deps.nix

```diff
@@ -0,0 +1,154 @@
+    SRC="$(mktemp -d)"
+    trap "rm -rf '$SRC'" EXIT
+    cp -r "$source/src" "$SRC/"
+    chmod -R +w "$SRC/src"
+    cd "$SRC/src"
+    source "$source/script2"
+    ${preBuild}
+    ${gradleCommands}
+    ${postBuild}
+  '';
+in
+writeShellScript "fetch-deps.sh" ''
+  set -eo pipefail
+  ${lib.optionalString (lib.hasInfix "$attrPath" code) ''
+    attrPath=${lib.escapeShellArg attrPath}
+    [[ -n "$attrPath" ]] || attrPath="$1"
+    [[ -n "$attrPath" ]] || attrPath="$UPDATE_NIX_ATTR_PATH"
+    [[ -n "$attrPath" ]] || (echo "Set UPDATE_NIX_ATTR_PATH or pass attr path as the first argument!" && exit 1)
+  ''}
+  outPath="${
+    # if this is an absolute path in nix store, use path relative to the output root
+    if lib.hasPrefix "/nix/store/" (toString depsPath)
+    then builtins.concatStringsSep "/" (lib.drop 4 (lib.splitString "/" (toString depsPath)))
+    # if this is an absolute path, just use that path
+    else if lib.hasPrefix "/" (toString depsPath)
+    then depsPath
+    # otherwise, use a path relative to the package
+    else ''$(nix-instantiate --eval -E "dirOf (with (${code}); meta.position)" | tr -d '"')/${depsPath}''
+  }"
+
+  source="$(nix-build --no-out-link -E "import ${./patched-source.nix} ({ self = ${code}; } // builtins.fromJSON "${
+    lib.escapeShellArg (builtins.toJSON (builtins.toJSON attrOverrides))
+  }")")"
+
+  source "$source/script1"
+  pushd "$(mktemp -d)"
+  MITM_CACHE_DIR="$PWD"
+  trap "rm -rf '$MITM_CACHE_DIR'" EXIT
+  ${openssl}/bin/openssl genrsa -out ca.key 2048
+  ${openssl}/bin/openssl req -x509 -new -nodes -key ca.key -sha256 -days 1 -out ca.cer -subj "/C=AL/ST=a/L=a/O=a/OU=a/CN=example.org"
+  export MITM_CACHE_PORT="''${mitmCachePort:-${toString mitmCachePort}}"
+  ${mitm-cache}/bin/mitm-cache \
+    -l"127.0.0.1:$MITM_CACHE_PORT" \
+    record \
+    --reject '\.(md5|sha(1|256|512:?):?)$' \
+    --forget-redirects-from '://(download\.savannah\.gnu\.org|github\.com|packages\.jetbrains\.team|sourceforge\.net:?)/' \
+    --forget-redirects-to '://(cdn\.jsdelivr\.net|plugins-artifacts\.gradle\.org:?)/' \
+    --record-text '/maven-metadata\.xml$' >/dev/null 2>/dev/null &
+  MITM_CACHE_PID="$!"
+  trap "kill '$MITM_CACHE_PID'" EXIT
+  sleep 0.5
+  ps -p "$MITM_CACHE_PID" >/dev/null || (echo "Failed to launch mitm-cache! Port $MITM_CACHE_PORT may be already used. Try setting "'$mitmCachePort' && exit 1)
+  export MITM_CACHE_CA="$PWD/ca.cer"
+  export MITM_CACHE_KS_PWD="$(head -c10 /dev/random | base32)"
+  echo y | ${jdk}/bin/keytool -importcert -file ca.cer -alias alias -keystore keystore -storepass "$MITM_CACHE_KS_PWD"
+  export MITM_CACHE_KEYSTORE="$PWD/keystore"
+  popd
+  ${preBuildNoSandbox}
+  # nsjail isn't necessary, it's only used to prevent messy build scripts from touching ~
+  ${nsjail}/bin/nsjail \
+    -N --keep_caps --chroot / --disable_rlimits \
+    -T /home -T /root -R ${path} \
+    ${lib.escapeShellArgs (map (x: "-E${x}") keep)} \
+    -E NIX_PATH=nixpkgs=${path} \
```

Comment:
Somehow this path is forced to be flake-based? It always copies Nixpkgs to the Nix store, even when building without flakes which basically changes the fetcher script it every time as the result symlink always changes.

Reply:
interesting, in all-packages.nix it's defined as simply `../..`

Reply:
Ah, in that case it's obvious. String-substituting a path always results in a copy in the Nix store; trivial to reproduce using `"${./.}"` in a repl.

The fix should be to `toString` it; i.e. `"${toString ./.}"`. Wondering whether that should be the default for `pkgs.path` and I'm leaning towards yes.

Reply:
cc @roberth

Reply:
`toString` is not guaranteed to produce a path in the store, so an unconditional `toString` only works for flakes, and even that might change.

Here's my attempt to fix `pkgs.path`:
- https://github.com/NixOS/nixpkgs/pull/153594

It's complicated, and the problems we hit there will hit this too.

Do we really need to access Nixpkgs here?

Reply:
I can try to sidestep this by getting the `pname` attr location using the undocumented builtin

Reply:
either way, i'm using toString now